src/data/program_configuration.js

```diff
@@ -371,8 +364,8 @@ class CrossFadedCompositeBinder implements AttributeBinder {
 
     upload(context: Context) {
         if (this.zoomInPaintVertexArray && this.zoomInPaintVertexArray.arrayBuffer && this.zoomOutPaintVertexArray && this.zoomOutPaintVertexArray.arrayBuffer) {
-            this.zoomInPaintVertexBuffer = context.createVertexBuffer(this.zoomInPaintVertexArray, this.paintVertexAttributes, this.expression.isStateDependent);
-            this.zoomOutPaintVertexBuffer = context.createVertexBuffer(this.zoomOutPaintVertexArray, this.paintVertexAttributes, this.expression.isStateDependent);
+            this.zoomInPaintVertexBuffer = context.createVertexBuffer(this.zoomInPaintVertexArray, patternAttributes.members, this.expression.isStateDependent);
+            this.zoomOutPaintVertexBuffer = context.createVertexBuffer(this.zoomOutPaintVertexArray, patternAttributes.members, this.expression.isStateDependent);
```

Comment:
Isn't this hard-coding though? are patterns the only kind of corss faded composite attributes?

Reply:
Yes they are the only type of vertices at this time, we don't emplace any other type in this layout here so we can make it explicit https://github.com/mapbox/mapbox-gl-js/pull/9482/files#diff-0ecc9a72f8fbadd5fe4d9a5f4ab26691R350. We can leave a note to prevent programmer error if this gets extended.

Reply:
I think the note should be in `paintAttributeNames` [here](https://github.com/mapbox/mapbox-gl-js/blob/b238b15f64e00540c29d0ec3d2065861aa90d1f9/src/data/program_configuration.js#L625), as adding attributes might lead to issues with the other type of binders such as [here](https://github.com/mapbox/mapbox-gl-js/blob/b238b15f64e00540c29d0ec3d2065861aa90d1f9/src/data/program_configuration.js#L241) or [here](https://github.com/mapbox/mapbox-gl-js/blob/b238b15f64e00540c29d0ec3d2065861aa90d1f9/src/data/program_configuration.js#L170).

Reply:
How about cross-referencing from a lookup table?

We could either update the lookup table by hand, or codegen it in `generate-struct-arrays.js`.

Reply:
Logging offline discussion: we decided on going the simple way for the time being to cut out extra logic.

Reply:
👍